tests/test_quaternion.h

```diff
@@ -0,0 +1,366 @@
+	CHECK(q_r[3] == doctest::Approx(0.996195));
+
+	// Test ZYX dynamic-axes since test data is available online.
+	//  Rotate first about X axis, then new Y axis, then new Z axis.
+	// (Godot uses YXZ Yaw-Pitch-Roll order).
+	Quaternion q_yp = q_y * q_p;
+	CHECK(q_yp[0] == doctest::Approx(0.239118));
+	CHECK(q_yp[1] == doctest::Approx(0.369644));
+	CHECK(q_yp[2] == doctest::Approx(-0.099046));
+	CHECK(q_yp[3] == doctest::Approx(0.892399));
+
+	Quaternion q_ryp = q_r * q_yp;
+	CHECK(q_ryp[0] == doctest::Approx(0.205991));
+	CHECK(q_ryp[1] == doctest::Approx(0.389078));
+	CHECK(q_ryp[2] == doctest::Approx(-0.0208912));
+	CHECK(q_ryp[3] == doctest::Approx(0.897636));
+}
+
+TEST_CASE("[Quaternion] xform unit vectors") {
+	// Easy to visualize: 120 deg about X-axis.
+	// Transform the i, j, & k unit vectors.
+	Quaternion q(Vector3(1.0, 0.0, 0.0), deg2rad(120.0));
+	Vector3 i_t = q.xform(Vector3(1.0, 0.0, 0.0));
+	Vector3 j_t = q.xform(Vector3(0.0, 1.0, 0.0));
+	Vector3 k_t = q.xform(Vector3(0.0, 0.0, 1.0));
+	//
+	CHECK(i_t.is_equal_approx(Vector3(1.0, 0.0, 0.0)));
+	CHECK(j_t.is_equal_approx(Vector3(0.0, -0.5, 0.866025)));
+	CHECK(k_t.is_equal_approx(Vector3(0.0, -0.866025, -0.5)));
+	CHECK(i_t.length_squared() == doctest::Approx(1.0));
+	CHECK(j_t.length_squared() == doctest::Approx(1.0));
+	CHECK(k_t.length_squared() == doctest::Approx(1.0));
+
+	// Easy to visualize: 30 deg about Y-axis.
+	q = Quaternion(Vector3(0.0, 1.0, 0.0), deg2rad(30.0));
+	i_t = q.xform(Vector3(1.0, 0.0, 0.0));
+	j_t = q.xform(Vector3(0.0, 1.0, 0.0));
+	k_t = q.xform(Vector3(0.0, 0.0, 1.0));
+	//
+	CHECK(i_t.is_equal_approx(Vector3(0.866025, 0.0, -0.5)));
+	CHECK(j_t.is_equal_approx(Vector3(0.0, 1.0, 0.0)));
+	CHECK(k_t.is_equal_approx(Vector3(0.5, 0.0, 0.866025)));
+	CHECK(i_t.length_squared() == doctest::Approx(1.0));
+	CHECK(j_t.length_squared() == doctest::Approx(1.0));
+	CHECK(k_t.length_squared() == doctest::Approx(1.0));
+
+	// Easy to visualize: 60 deg about Z-axis.
+	q = Quaternion(Vector3(0.0, 0.0, 1.0), deg2rad(60.0));
+	i_t = q.xform(Vector3(1.0, 0.0, 0.0));
+	j_t = q.xform(Vector3(0.0, 1.0, 0.0));
+	k_t = q.xform(Vector3(0.0, 0.0, 1.0));
+	//
+	CHECK(i_t.is_equal_approx(Vector3(0.5, 0.866025, 0.0)));
+	CHECK(j_t.is_equal_approx(Vector3(-0.866025, 0.5, 0.0)));
+	CHECK(k_t.is_equal_approx(Vector3(0.0, 0.0, 1.0)));
+	CHECK(i_t.length_squared() == doctest::Approx(1.0));
+	CHECK(j_t.length_squared() == doctest::Approx(1.0));
+	CHECK(k_t.length_squared() == doctest::Approx(1.0));
+}
+
+TEST_CASE("[Quaternion] xform vector") {
+	// Arbitrary quaternion rotates an arbitrary vector.
+	Vector3 euler_yzx(deg2rad(31.41), deg2rad(-49.16), deg2rad(12.34));
+	Basis basis_axes(euler_yzx);
+	Quaternion q(basis_axes);
+
+	Vector3 v_arb(3.0, 4.0, 5.0);
+	Vector3 v_rot = q.xform(v_arb);
+	Vector3 v_compare = basis_axes.xform(v_arb);
+	//
+	CHECK(v_rot.length_squared() == doctest::Approx(v_arb.length_squared()));
+	CHECK(v_rot.is_equal_approx(v_compare));
+}
+
```

Comment:
Do you want to do stress test generate large numbers of quaternions? See basis tests for examples.

Reply:
Here's a good-faith test case that exercises many quaternions & vectors. I didn't include random numbers since that would make the test non-repeatable. I don't think I fully grasp the project approach to stress tests, but the new test case will exercise a much wider range of values.